File: tests/Chessie.CSharp.Test/SimpleValidation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using NUnit.Framework;
using Chessie.ErrorHandling;
using Chessie.ErrorHandling.CSharp;

namespace Chessie.CSharp.Test
{
    public class Request
    {
        public string Name { get; set; }
        public string EMail { get; set; }
    }

    public class Validation
    {
        public static Result<Request, string> ValidateInput(Request input)
        {
            if (input.Name == "")
                return Result<Request, string>.FailWith("Name must not be blank");
            if (input.EMail == "")
                return Result<Request, string>.FailWith("Email must not be blank");
            return Result<Request, string>.Succeed(input);

        }
    }

    [TestFixture]
    public class TrySpecs
    {
        [Test]
        public void TryWillCatch()
        {
            var exn = new Exception("Hello World");
            var result = Result<string, Exception>.Try(() => { throw exn; });
            Assert.AreEqual(exn, result.FailedWith().First());
        }

        [Test]
        public void TryWillReturnValue()
        {
            var result = Result<string, Exception>.Try(() => { return "hello world"; });
            Assert.AreEqual("hello world", result.SucceededWith());
        }
    }

    [TestFixture]
    public class SimpleValidation
    {
        [Test]
        public void CanCreateSuccess()
        {
            var request = new Request { Name = "Steffen", EMail = "mail@support.com" };
            var result = Validation.ValidateInput(request);
            Assert.AreEqual(request, result.SucceededWith());
        }
    }

    [TestFixture]
    public class SimplePatternMatching
    {
        [Test]
        public void CanMatchSuccess()
        {
            var request = new Request { Name = "Steffen", EMail = "mail@support.com" };
            var result = Validation.ValidateInput(request);
            result.Match(
               (x, msgs) => { Assert.AreEqual(request, x); },
               msgs => { throw new Exception("wrong match case"); });
        }

        [Test]
        public void CanMatchFailure()
        {
            var request = new Request { Name = "Steffen", EMail = "" };
            var result = Validation.ValidateInput(request);
            result.Match(
               (x, msgs) => { throw new Exception("wrong match case"); },
               msgs => { Assert.AreEqual("Email must not be blank", msgs[0]); });
        }
    }

    [TestFixture]
    public class SimpleEitherPatternMatching
    {
        [Test]
        public void CanMatchSuccess()
        {
            var request = new Request { Name = "Steffen", EMail = "mail@support.com" };
            var result =
                Validation
                 .ValidateInput(request)
                 .Either(
                   (x, msgs) => { return x; },
                   msgs => { throw new Exception("wrong match case"); });
            Assert.AreEqual(request, result);
        }

        [Test]
        public void CanMatchFailure()
        {
            var request = new Request { Name = "Steffen", EMail = "" };
            var result =
               Validation.ValidateInput(request)
                .Either(
                   (x, msgs) => { throw new Exception("wrong match case"); },
                   msgs => { return msgs[0]; });

            Assert.AreEqual("Email must not be blank", result);
        }
    }
}

File: tests/Chessie.VisualBasic.Tests/SimpleValidation.vb
Imports Chessie.ErrorHandling
Imports Chessie.ErrorHandling.CSharp
Imports NUnit.Framework

Public Class Request
    Public Property Name() As String
    Public Property EMail() As String
End Class

Public Class Validation
    Public Shared Function ValidateInput(input As Request) As Result(Of Request, String)
        If input.Name = "" Then
            Return Result(Of Request, String).FailWith("Name must not be blank")
        End If
        If input.EMail = "" Then
            Return Result(Of Request, String).FailWith("Email must not be blank")
        End If
        Return Result(Of Request, String).Succeed(input)

    End Function
End Class

<TestFixture>
Public Class TrySpecs
    <Test>
    Public Sub TryWillCatch()
        Dim exn = New Exception("Hello World")
        Dim result = ErrorHandling.Result(Of String, Exception).[Try](Function()
                                                                          Throw exn
                                                                          Return ""
                                                                      End Function)
        Assert.AreEqual(exn, result.FailedWith().First())
    End Sub

    <Test>
    Public Sub TryWillReturnValue()
        Dim result = ErrorHandling.Result(Of String, Exception).[Try](Function() "hello world")
        Assert.AreEqual("hello world", result.SucceededWith())
    End Sub
End Class

<TestFixture>
Public Class SimpleValidation
    <Test>
    Public Sub CanCreateSuccess()
        Dim request = New Request() With {
            .Name = "Steffen",
            .EMail = "mail@support.com"
        }
        Dim result = Validation.ValidateInput(request)
        Assert.AreEqual(request, result.SucceededWith())
    End Sub
End Class

<TestFixture>
Public Class SimplePatternMatching
    <Test>
    Public Sub CanMatchSuccess()
        Dim request = New Request() With {
            .Name = "Steffen",
            .EMail = "mail@support.com"
        }
        Dim result = Validation.ValidateInput(request)
        result.Match(Sub(x, msgs) Assert.AreEqual(request, x),
                     Sub(msgs) Throw New Exception("wrong match case"))
    End Sub

    <Test>
    Public Sub CanMatchFailure()
        Dim request = New Request() With {
            .Name = "Steffen",
            .EMail = ""
        }
        Dim result = Validation.ValidateInput(request)
        result.Match(Sub(x, msgs) Throw New Exception("wrong match case"),
                     Sub(msgs) Assert.AreEqual("Email must not be blank", msgs(0)))
    End Sub
End Class

<TestFixture>
Public Class SimpleEitherPatternMatching
    <Test>
    Public Sub CanMatchSuccess()
        Dim request = New Request() With {
            .Name = "Steffen",
            .EMail = "mail@support.com"
        }
        Dim result = Validation.ValidateInput(request).Either(Function(x, msgs) x,
                                                              Function(msgs)
                                                                  Throw New Exception("wrong match case")
                                                                  Return request
                                                              End Function)
        Assert.AreEqual(request, result)
    End Sub

    <Test>
    Public Sub CanMatchFailure()
        Dim request = New Request() With {
            .Name = "Steffen",
            .EMail = ""
        }
        Dim result = Validation.ValidateInput(request).Either(Function(x, msgs)
                                                                  Throw New Exception("wrong match case")
                                                                  Return ""
                                                              End Function,
                                                              Function(msgs) msgs(0))

        Assert.AreEqual("Email must not be blank", result)
    End Sub
End Class
